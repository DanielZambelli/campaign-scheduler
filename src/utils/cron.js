const CronJob = require('cron').CronJob

const cron = (cronStr, callback, autoStart = true) => {
  return new CronJob(cronStr, callback ,null, autoStart)
}

const cronOnce = (cronStr, callback, autoStart = true) => {
  const job = cron(cronStr, async () => {
    await callback()
    job.stop()
  }, autoStart)
  return job
}

module.exports = {cron,cronOnce}
