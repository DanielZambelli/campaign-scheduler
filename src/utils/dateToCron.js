const dateToCron = (dateString) => {
  const date = new Date(dateString)
  const seconds = date.getSeconds()
  const minutes = date.getMinutes()
  const hours = date.getHours()
  const days = date.getDate()
  const months = date.getMonth()
  const dayOfWeek = date.getDay()
  return `${seconds} ${minutes} ${hours} ${days} ${months} ${dayOfWeek}`
}

module.exports = {dateToCron}
