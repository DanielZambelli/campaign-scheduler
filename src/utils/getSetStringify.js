const getSetStringify = (field) => {
  return {
    get(){
      const value = this.getDataValue(field)
      return value ? JSON.parse(value) : value
    },
    set(value){
      if(value) value = JSON.stringify(value)
      this.setDataValue(field, value)
    }
  }
}

module.exports = {
  getSetStringify
}
