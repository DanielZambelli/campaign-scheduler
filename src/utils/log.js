const util = require('util')
const log = (object) => console.log(util.inspect(object, false, null, true))

module.exports = {log}
