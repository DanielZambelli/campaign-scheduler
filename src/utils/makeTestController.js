const { CampaignScheduler: Controller } = require('../controller/index')

const makeTestController = (id, callbacks = {}) => {

  const db = process.env.CS_DB ? JSON.parse(process.env.CS_DB) : { "dialect": "sqlite" }
  if(db.dialect==='sqlite') db.storage = `tmp/db_${id}.sqlite`

  db.schema = `test_${id}`
  db.force = true
  db.makeDestroy = true

  return new Controller({ db, callbacks }).init()
}

module.exports = {
  makeTestController
}
