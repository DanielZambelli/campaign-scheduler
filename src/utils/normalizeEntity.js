const normalizeEntity = (entity) => {
  const {createdAt,updatedAt,...entityNormalized} = entity.toJSON()
  return entityNormalized
}

module.exports = {
  normalizeEntity
}
