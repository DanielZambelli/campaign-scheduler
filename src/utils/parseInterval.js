const moment = require('moment')
const {parseIntervalOffset} = require('./parseIntervalOffset')
const {parseIntervalTime} = require('./parseIntervalTime')

const parseInterval = (interval, startDate=undefined) => {

  const {offset=null, time=null} = interval
  startDate = moment(startDate)

  //start from now or offset from previous date
  const date = moment(startDate)

  // offset
  const offsetParsed = parseIntervalOffset(offset)
  if(offsetParsed) date.add(offsetParsed)

  // specific time
  const timeParsed = parseIntervalTime(time)
  if(timeParsed) date.set(timeParsed)

  // incase now has passed the time slot offset till next day
  if(date.isSame(startDate, 'day') && startDate.isAfter(date))
    date.add({ days: 1 })

  return date.toISOString()
}

module.exports = {parseInterval}
