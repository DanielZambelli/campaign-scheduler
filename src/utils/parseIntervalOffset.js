const {getRandomBetween} = require('./getRandomBetween')

const parseIntervalOffset = (offset) => {
  if(!offset) return
  let [a,b] = offset.split('-').map(offset => {
    return offset.trim().split(' ').reduce((map,item) => {
      const value = parseInt(item.match(/\d+/g))
      const unit = item.replace(value,'')
      map[unit] = value
      return map
    }, {})
  })
  if(a && b){
    Object.keys(b).forEach(unit => {
      if(a[unit]){
        const min = Math.min(a[unit], b[unit])
        const max = Math.max(a[unit], b[unit])
        a[unit] = getRandomBetween(min,max)
      }else a[unit] = b[unit]
    })
  }
  return a
}

module.exports = {parseIntervalOffset}
