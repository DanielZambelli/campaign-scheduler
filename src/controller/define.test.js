require('../utils/test')
const {seed} = require('../utils/test/seeds1')
const TEST = 'define'

const action1 = { id: 'action1', interval: { offset: '1hours' }, callback: { id: 'sendEmail', opts: { view: 'template1' } } }
const action2 = { id: 'action2', interval: { offset: '2hours' }, callback: { id: 'sendEmail', opts: { view: 'template2' } } }

const smsActions = [
  { id: 'action1', interval: { offset: '1day' },  callback: { id: 'sendSms', view: 'smsTemplate1' } },
  { id: 'action2', interval: { offset: '2days' }, callback: { id: 'sendSms', view: 'smsTemplate2' } },
  { id: 'action3', interval: { offset: '3days' }, callback: { id: 'sendSms', view: 'smsTemplate3' } },
]

describe(TEST, () => {

  beforeEach(init(TEST,seed))

  afterEach(destroy)

  it('validates', async () => {
    const res = await ctl.define({ id: undefined, actions: [] }).catch(e => e.message)
    expect(res).toEqual('notNull Violation: cs_campaigns.id cannot be null')
  })

  it('validates', async () => {
    const res = await ctl.define({ id: 'myCampaign', actions: [ { id: 'a1' } ] }).catch(e => e.message)
    expect(res).toEqual('Validation error: actions item requires options')
  })

  it('validates', async () => {
    const res = await ctl.define({ id: 'myCampaign', actions: [ { id: 'a1', interval: null, callback: { id: 'sendEmail' } } ]}).catch(e => e.message)
    expect(res).toEqual('Validation error: actions item requires options')
  })

  it('validates', async () => {
    const res = await ctl.define({ id: 'myCampaign', actions: [ { id: 'a1', interval: { offset: '5hours' }, callback: null } ]}).catch(e => e.message)
    expect(res).toEqual('Validation error: actions item requires options')
  })

  it('returns', async () => {
    const res = await ctl.define({ id: 'myCampaign', active: true, actions: [ action1 ]}).then(normalize)
    expect(res.result.actions.length).toEqual(1)
    expect(res).toMatchSnapshot()
  })

  it('returns', async () => {
    const res = await ctl.define({ id: 'myCampaign', active: false, actions: [ action1 ]}).then(normalize)
    expect(res.result.actions.length).toEqual(1)
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    const campaign = { id: 'myCampaign', active: true, actions: [ action1 ]}
    await ctl.define(campaign)
    const res = await getState({ campaignId: campaign.id })
    expect(res.state.campaigns[0]).toEqual(campaign)
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    const campaign = { id: 'myCampaign', active: false, actions: [ action2 ]}
    await ctl.define(campaign)
    const res = await getState({ campaignId: campaign.id })
    expect(res.state.campaigns[0]).toEqual(campaign)
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    const campaign = { id: 'myCampaign', active: true, actions: [ action1, action2 ]}
    await ctl.define(campaign)
    const res = await getState({ campaignId: campaign.id })
    expect(res.state.campaigns[0]).toEqual(campaign)
    expect(res).toMatchSnapshot()
  })

  it('state', async () => {
    const campaign = { id: 'myCampaign', active: false, actions: [ action1, action2 ]}
    await ctl.define(campaign)
    const res = await getState({ campaignId: campaign.id })
    expect(res.state.campaigns[0]).toEqual(campaign)
    expect(res).toMatchSnapshot()
  })

  it('reconciliates active', async () => {
    await ctl.schedule('campaign3', 'contact#1')
    await ctl.schedule('campaign3', 'contact#2')
    await ctl.db.Actions.update({ state: 'completed' }, { where: { campaignId: 'campaign3', actionId: 'action1', subject: ['contact#1','contact#2'] } } )
    await ctl.define({ id: 'campaign3', active: true, actions: [...smsActions] })
    const res = await getState({ campaignId: 'campaign3' })
    expect(res).toMatchSnapshot()
  })

  it('reconciliates inactivated', async () => {
    await ctl.schedule('campaign1', 'contact#1')
    await ctl.schedule('campaign1', 'contact#2')
    await ctl.db.Actions.update({ state: 'completed' }, { where: { campaignId: 'campaign1', actionId: 'action1', subject: ['contact#2'] } } )
    await ctl.define({ id: 'campaign1', active: false, actions: [...smsActions] })
    const res = await getState({ campaignId: 'campaign1' })
    expect(res.state.campaigns[0].active).toEqual(false)
    expect(res.state.actions.length).toEqual(1)
    expect(res.state.actions[0].state).toEqual('completed')
    expect(res).toMatchSnapshot()
  })

  it('reconciliates activated', async () => {
    await ctl.schedule('campaign2', 'contact#1')
    await ctl.schedule('campaign2', 'contact#2')
    await ctl.schedule('campaign2', 'contact#3')
    await ctl.unschedule('campaign2', 'contact#3')
    await ctl.define({
      id: 'campaign2', active: true, actions: [
        { id: 'action1', interval: { offset: '1day' }, callback: { id: 'sendSms', view: 'smsTemplate1' } },
        { id: 'action2', interval: { offset: '2days' }, callback: { id: 'sendSms', view: 'smsTemplate2' } },
      ],
    })
    const res = await getState({campaignId: 'campaign2' })
    expect(res.state.campaigns[0].active).toEqual(true)
    expect(res.state.actions.length).toEqual(4)
    expect(res).toMatchSnapshot()
  })

  it('reconciliates inactive', async () => {
    await ctl.schedule('campaign2', 'contact#1')
    await ctl.schedule('campaign2', 'contact#2')
    await ctl.unschedule('campaign2', 'contact#2')
    await ctl.define({ id: 'campaign2', active: false, actions: [
      { id: 'action3', interval: { offset: '0hours' }, callback: { id: 'sendEmail', opts: { view: 'template1' } } }
    ]})
    const res = await getState()
    expect(res.state.schedules[0].active).toEqual(true)
    expect(res.state.schedules[1].active).toEqual(false)
    expect(res.state.actions.length).toEqual(0)
    expect(res.state.campaigns.find(o => o.id === 'campaign2').active).toEqual(false)
    expect(res).toMatchSnapshot()
  })

})
