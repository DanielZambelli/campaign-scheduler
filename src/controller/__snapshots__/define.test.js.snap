// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`define reconciliates activated 1`] = `
{
  "hash": "32f8ceee2fc83217c055917125903c96b0ea0d5d",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendSms",
        "callbackOpts": {
          "view": "smsTemplate1",
        },
        "campaignId": "campaign2",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
      {
        "actionId": "action2",
        "callbackId": "sendSms",
        "callbackOpts": {
          "view": "smsTemplate2",
        },
        "campaignId": "campaign2",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
      {
        "actionId": "action1",
        "callbackId": "sendSms",
        "callbackOpts": {
          "view": "smsTemplate1",
        },
        "campaignId": "campaign2",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "pending",
        "subject": "contact#2",
      },
      {
        "actionId": "action2",
        "callbackId": "sendSms",
        "callbackOpts": {
          "view": "smsTemplate2",
        },
        "campaignId": "campaign2",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#2",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendSms",
              "view": "smsTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1day",
            },
          },
          {
            "callback": {
              "id": "sendSms",
              "view": "smsTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign2",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#1",
      },
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#2",
      },
      {
        "active": false,
        "campaignId": "campaign2",
        "subject": "contact#3",
      },
    ],
  },
}
`;

exports[`define reconciliates active 1`] = `
{
  "hash": "bbdb9c4047a7714e06187bb94babf0b5aac9ec01",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "completed",
        "subject": "contact#1",
      },
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in a day",
        "log": null,
        "state": "completed",
        "subject": "contact#2",
      },
      {
        "actionId": "action2",
        "callbackId": "sendSms",
        "callbackOpts": {
          "view": "smsTemplate2",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
      {
        "actionId": "action3",
        "callbackId": "sendSms",
        "callbackOpts": {
          "view": "smsTemplate3",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 3 days",
        "log": null,
        "state": "pending",
        "subject": "contact#1",
      },
      {
        "actionId": "action2",
        "callbackId": "sendSms",
        "callbackOpts": {
          "view": "smsTemplate2",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 2 days",
        "log": null,
        "state": "pending",
        "subject": "contact#2",
      },
      {
        "actionId": "action3",
        "callbackId": "sendSms",
        "callbackOpts": {
          "view": "smsTemplate3",
        },
        "campaignId": "campaign3",
        "completedAt": null,
        "expectedAt": "in 3 days",
        "log": null,
        "state": "pending",
        "subject": "contact#2",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendSms",
              "view": "smsTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1day",
            },
          },
          {
            "callback": {
              "id": "sendSms",
              "view": "smsTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
          {
            "callback": {
              "id": "sendSms",
              "view": "smsTemplate3",
            },
            "id": "action3",
            "interval": {
              "offset": "3days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#1",
      },
      {
        "active": true,
        "campaignId": "campaign3",
        "subject": "contact#2",
      },
    ],
  },
}
`;

exports[`define reconciliates inactivated 1`] = `
{
  "hash": "6c96ce4db1f193b0cfc14ef3c904481212a3fff6",
  "state": {
    "actions": [
      {
        "actionId": "action1",
        "callbackId": "sendEmail",
        "callbackOpts": {
          "view": "emailTemplate1",
        },
        "campaignId": "campaign1",
        "completedAt": null,
        "expectedAt": "in an hour",
        "log": null,
        "state": "completed",
        "subject": "contact#2",
      },
    ],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendSms",
              "view": "smsTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1day",
            },
          },
          {
            "callback": {
              "id": "sendSms",
              "view": "smsTemplate2",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
          {
            "callback": {
              "id": "sendSms",
              "view": "smsTemplate3",
            },
            "id": "action3",
            "interval": {
              "offset": "3days",
            },
          },
        ],
        "active": false,
        "id": "campaign1",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#1",
      },
      {
        "active": true,
        "campaignId": "campaign1",
        "subject": "contact#2",
      },
    ],
  },
}
`;

exports[`define reconciliates inactive 1`] = `
{
  "hash": "fc7176698a346d5d7a5515541089503af7c6c827",
  "state": {
    "actions": [],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "campaign1",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "opts": {
                "view": "template1",
              },
            },
            "id": "action3",
            "interval": {
              "offset": "0hours",
            },
          },
        ],
        "active": false,
        "id": "campaign2",
      },
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action1",
            "interval": {
              "offset": "1days",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "view": "emailTemplate1",
            },
            "id": "action2",
            "interval": {
              "offset": "2days",
            },
          },
        ],
        "active": true,
        "id": "campaign3",
      },
    ],
    "schedules": [
      {
        "active": true,
        "campaignId": "campaign2",
        "subject": "contact#1",
      },
      {
        "active": false,
        "campaignId": "campaign2",
        "subject": "contact#2",
      },
    ],
  },
}
`;

exports[`define returns 1`] = `
{
  "hash": "78ffda5241ce7e819be150994e940b208769a01f",
  "result": {
    "actions": [
      {
        "callback": {
          "id": "sendEmail",
          "opts": {
            "view": "template1",
          },
        },
        "id": "action1",
        "interval": {
          "offset": "1hours",
        },
      },
    ],
    "active": true,
    "createdAt": "a few seconds ago",
    "id": "myCampaign",
    "updatedAt": "a few seconds ago",
  },
}
`;

exports[`define returns 2`] = `
{
  "hash": "849fad3564881defe576d1cb94bee2c140cd6d6c",
  "result": {
    "actions": [
      {
        "callback": {
          "id": "sendEmail",
          "opts": {
            "view": "template1",
          },
        },
        "id": "action1",
        "interval": {
          "offset": "1hours",
        },
      },
    ],
    "active": false,
    "createdAt": "a few seconds ago",
    "id": "myCampaign",
    "updatedAt": "a few seconds ago",
  },
}
`;

exports[`define state 1`] = `
{
  "hash": "b0bb5da67bec816e6bbaba8f6deecac3388a7867",
  "state": {
    "actions": [],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "opts": {
                "view": "template1",
              },
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
        ],
        "active": true,
        "id": "myCampaign",
      },
    ],
    "schedules": [],
  },
}
`;

exports[`define state 2`] = `
{
  "hash": "9b69133e10e5a97c65a6f5a8c56e28a0daaf5af5",
  "state": {
    "actions": [],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "opts": {
                "view": "template2",
              },
            },
            "id": "action2",
            "interval": {
              "offset": "2hours",
            },
          },
        ],
        "active": false,
        "id": "myCampaign",
      },
    ],
    "schedules": [],
  },
}
`;

exports[`define state 3`] = `
{
  "hash": "cd17c1d99214cca9bd6dddbcc4b97d3caef7a800",
  "state": {
    "actions": [],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "opts": {
                "view": "template1",
              },
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "opts": {
                "view": "template2",
              },
            },
            "id": "action2",
            "interval": {
              "offset": "2hours",
            },
          },
        ],
        "active": true,
        "id": "myCampaign",
      },
    ],
    "schedules": [],
  },
}
`;

exports[`define state 4`] = `
{
  "hash": "1096469a6a88f0627efb8593a8a1d30886484780",
  "state": {
    "actions": [],
    "campaigns": [
      {
        "actions": [
          {
            "callback": {
              "id": "sendEmail",
              "opts": {
                "view": "template1",
              },
            },
            "id": "action1",
            "interval": {
              "offset": "1hours",
            },
          },
          {
            "callback": {
              "id": "sendEmail",
              "opts": {
                "view": "template2",
              },
            },
            "id": "action2",
            "interval": {
              "offset": "2hours",
            },
          },
        ],
        "active": false,
        "id": "myCampaign",
      },
    ],
    "schedules": [],
  },
}
`;
